Add support for the Banana Pi BPI-F3

akifejaz opened the pull request:
This PR introduces a new seL4 platform port for the [Banana Pi BPI-F3](https://docs.banana-pi.org/en/BPI-F3/BananaPi_BPI-F3) development board, having the SpacemiT K1 RISC-V SoC.

The SpacemiT K1 is an 8 Core RV64GCVB/IMACFDCV board, supporting the RISC-V Vector extension (RVV 1.0). 

The device tree is extracted from the latest Bianbu OS [v3.0.1](https://archive.spacemit.com/image/k1/version/bianbu/v3.0.1/) also present at [linux-k1x](https://github.com/spacemit-com/linux-k1x/tree/k1/arch/riscv/boot/dts/spacemit), no other changes has been made in this dts.

Please review and let me know if needed anything else.

akifejaz commented on src/plat/spacemit-k1/config.cmake:
```diff
@@ -0,0 +1,29 @@
+    declare_seL4_arch(riscv64)
+    config_set(KernelRiscVPlatform RISCV_PLAT ${KernelPlatform})
+    config_set(KernelPlatformFirstHartID FIRST_HART_ID 0)
+    config_set(KernelOpenSBIPlatform OPENSBI_PLATFORM "generic")
```
I tried to mimic this config as per hifive-p550 [eswin/config.cmake ](https://github.com/seL4/seL4/blob/master/src/plat/eswin/config.cmake). 

but sure, I'll remove unused options.

Indanz commented:
tools/dts/bananapi-f3.dts has trailing whitespace on line 10 and Gitlint complains commit title length:

```
Checking the following commits:
d4a439ce Add support for the Banana Pi BPI-F3 board (Spacemit K1 SoC)

1: T1 Title exceeds max length (60>50): "Add support for the Banana Pi BPI-F3 board (Spacemit K1 SoC)"
```
You can add "Based on the Spacemit K1 SoC" in the commit message instead of the title.

Indanz commented:
> sorry, what went wrong, didnt understand. I removed the trailing space and commit the change, thats it.

I assumed you wanted one commit message with a title "Add support for the Banana Pi BPI-F3" and "Based on the Spacemit K1 SoC" in the body of the commit message. Now you have two commits and the first one still fails the PR check because the title is still too long.

Indanz commented:
> Yes, I'had plan to PR that, but I'm also testing existing timer driver (if that works), I need some time on that.

That's okay, either would be fine. It's always good to have reference drivers. If both work you can make it a config option which timer to use.

> Do we need that along with this PR or can we go with this PR for now?

sel4test and sel4bench support is the minimum for new platforms. Both are good tests to see if the port has been done correctly. If sel4test fails something is wrong. And if sel4bench is worse than expected, there may be some bug somewhere, or misconfiguration, or seL4 doing something that your hardware really doesn't like..

It would be nice to also get one or two boards to add to the CI system so it can be automatically tested. In this case it would also be handy to develop new features, like RVV support. This is really optional, but if you plan on using the hardware long-term, it's a very good investment to make sure seL4 keeps working for your hardware in the future, because any problems will be discovered early. If it's not part of the CI system, any future problems you have to fix and debug yourself. (Edit: You can always open an issue here and we will try to help of course, but for hands-on debugging you need access to the hardware itself.)

akifejaz commented:
> Hi, I would really like some help with timer driver I've added the driver changes in [seL4/util_libs#202](https://github.com/seL4/util_libs/pull/202) and facing [this issue](https://github.com/seL4/util_libs/pull/202#issuecomment-3479322089) also mentioned under same PR (I'll create issue later if needed)

Thanks all for help, the driver has been fixed and added here https://github.com/seL4/util_libs/pull/202

Indanz commented:
Do sel4test and sel4bench pass for all configurations? (SMP one and off, MCS on and off.)

Indanz commented:
You need to add your UART's compat string to tools/hardware.yml and the driver if it's missing (and also add a driver to the kernel if it's a new one).

akifejaz commented:
> BTW, is your seL4 printing working? I don't remember seeing you post any kernel output. 

Yes, I see prints from `Jumping to kernel-image entry point...` to `Booting all finished, dropped to user space` normally. I did share my logs in [here](https://github.com/seL4/util_libs/issues/203#issuecomment-3506684952) once.

akifejaz commented:
I can see the logs now, but it failed right away ... I'm not sure about the errors will look into it in details. 

```
Starting kernel ...

ELF-loader started on (HART 0) (NODES 1)
  paddr=[f77000..1850fff]
Looking for DTB in CPIO archive...found at f9c200.
Loaded DTB from f9c200.
   paddr=[a8000..bdfff]
ELF-loading image 'kernel' to 80000
  paddr=[80000..a7fff]
  vaddr=[ffffffff80080000..ffffffff800a7fff]
  virt_entry=ffffffff80080000
ELF-loading image 'rootserver' to be000
  paddr=[be000..9d6fff]
  vaddr=[10000..928fff]
  virt_entry=53c18
Enabling MMU and paging
Jumping to kernel-image entry point...

Init local IRQ
Bootstrapping kernel
Initializing PLIC...
available phys memory regions: 4
  [80000..100000)
  [600000..2ff40000)
  [30000000..7f000000)
  [100000000..480000000)
reserved virt address space regions: 3
  [ffffffc000080000..ffffffc0000a8000)
  [ffffffc0000a8000..ffffffc0000bdbd9)
  [ffffffc0000be000..ffffffc0009d7000)
Booting all finished, dropped to user space

[;1m[37mse[32mL4[37m Benchmark[0m
[32m==============[37m

Switching to a safer, bigger stack... 
Setting up global fault handler...
vka_alloc_object_at_maybe_dev@object.h:57 Failed to allocate object of size 274877906944, error 1
vka_alloc_object_at_maybe_dev@object.h:57 Failed to allocate object of size 137438953472, error 1
vka_alloc_object_at_maybe_dev@object.h:57 Failed to allocate object of size 68719476736, error 1
vka_alloc_object_at_maybe_dev@object.h:57 Failed to allocate object of size 34359738368, error 1
vka_alloc_object_at_maybe_dev@object.h:57 Failed to allocate object of size 17179869184, error 1

ipc Benchmarks
==============

libsel4muslcsys: Error attempting syscall 113
libsel4muslcsys: Error attempting syscall 169

irquser Benchmarks
==============


```

Indanz commented:
Ignore the "Failed to allocate", and probably the syscalls ones too. If it hangs on the irquser test, try disabling it (e.g. via ccmake) and see if others do work.

Indanz commented:
> Thanks for suggestion. I tried to disable but for some reason it always goes into the same path (little weird), like keep running IPC, IRQUSER and then hangs even after clean rebuild and double checking the ccmake. I think `libsel4muslcsys: Error attempting syscall 113` is error here and it sel4bench stops because of this and weirdly even disabling this, it still runs irq too

If you use `ccmake .` in the build directory, make sure you do both `c` to reconfigure it, and then `g` to generate the new config.

> btw I did not plan sel4bench I do not know its structure (little similar to sel4test though) but If I have to fix this at this point, I'll have to understand it.

High chance that it's unrelated to your port, it could also be a bug triggered by your port, or something else silly. I don't expect you to get to the bottom of this, just dig a bit till we have a better idea what's going on.

> one more thing, I've [Milk-V Jupiter](https://milkv.io/docs/jupiter/overview) board having the same SoC which I have plan to submit PR in coming week hopefully, so, I'll maybe try to work on sel4bench with that if it's not high priority at this point.

If it's the same SoC you should be able to re-use the same platform and just add it as a variant, with a different overlay.Not sure what the best way of doing that is in seL4, but there are some examples of that for other boards.

That's why it's better to use the SoC name for new platforms than the board name for this reason, which you did. That makes it much easier for people using a board with the same SoC to find the existing code.

akifejaz commented:
> High chance that it's unrelated to your port, it could also be a bug triggered by your port, or something else silly. I don't expect you to get to the bottom of this, just dig a bit till we have a better idea what's going on.

Sure, btw this PR needs rebase if you want to merge it. (I'll do shortly.)

> If it's the same SoC you should be able to re-use the same platform and just add it as a variant, with a different overlay.Not sure what the best way of doing that is in seL4, but there are some examples of that for other boards.
> 
> That's why it's better to use the SoC name for new platforms than the board name for this reason, which you did. That makes it much easier for people using a board with the same SoC to find the existing code.

yes, that is my plan too, and I've some work around that adds minimal patch for Jupiter and uses the same driver as by these PRs. Hopping to open PR in coming week.

midnightveil commented:
> I like this, and as I'm also a member of [RISC-V Ecosystem Labs](https://riscv.org/developers/labs/), I can setup CIs too for multiple RISC-V machines (as we've received multiple boards from RISC-V International). We have also setup automated system at [Cloud-V](https://cloud-v.co/) (project by [10xEngineers](https://10xengineers.ai/)) where we provide free access (SSH / GitHub CI / GitLab CI based access) to RISC-V Community.
> 
> I can surely setup some machines for SeL4.

@akifejaz We mostly have boards in physical hardware test racks at UNSW (Sydney). The way you phrased it, SSH or GitHub CI sounds more like software tests (e.g. code running on Linux)? Not sure that would be helpful for testing seL4... (And it would probably be extra effort to integrate our existing 'machine queue' with an external system).

lsf37 commented:
Sorry for the silence, Indan is away at the moment. He'll get back to you when he is online again.

Indanz commented:
> Hi @Indanz @lsf37, I would really appreciate if you could please proceed with this PR.
> 
> Let me know if needed any change from myside.

No, I did approve this PR, meaning the code is fine from my side. Main remaining issue was sel4bench, but that's apparently hopeless on newer RISC-V.

We also need to merge:
- https://github.com/seL4/seL4_tools/pull/240
- https://github.com/seL4/util_libs/pull/202

You need to fix the CI errors for the util_libs PR and it would be nice if you would squash the commits into one to clean up the history.

Once that's done, all is good to be merged from my side.

But it's a policy decision whether to accept all ports or not and the current policy is vague. I personally would, because that encourages people to use seL4 more, but I know Gerwin is more reluctant.

Of course we can't officially support the board if we don't have physical hardware in the seL4 CI, also meaning that if it ever gives maintenance problems we may remove it again, so that's the main caveat.

lsf37 commented:
> > sel4bench is not expected to work on newer RISC-V platforms because post 0.9 (I think it's that version but might be one before or one after) of OpenSBI prohibits user-space access to the PMU which is what partially motivates RFC-22. For my HiFive P550 port I ignored sel4bench for this reason.
> 
> Is this documented anywhere? Both on seL4's side and on RISC-V's side? That seems like a major decision that shouldn't be made by OpenSBI, but by the OS.

As soon as [RFC-22](https://github.com/seL4/rfcs/pull/35) is implemented, which hopefully is not too far away, we will require sel4bench for those RISC-V platforms again.

For now there appears to be nothing we can do for those boards, so we have to skip it.

> RISC-V seems to go the gutter very fast. What started as "let's implement microcode with M-mode" is turning into "let's do everything in M-mode".

It's certainly extremely annoying.
